### DISCUSSION TITLE
ad-hoc method for combining PRs

### DISCUSSION THREAD

**Discussion by dhduvall** [renovatebot/renovate#23310] [Suggest an Idea]
### Tell us more.

I know that you can configure renovate to combine multiple updates into a single PR, useful for when a bunch of dependencies are related in some way that makes this safe and sane.

But I would be interested in a way to do this on the fly. For instance, the initial set of PRs after onboarding for one of my repos were mostly trivial, and didn't need to land individually, triggering O(n^2) rebuilds as the rest got rebased. I ended up creating a new branch, merging renovate's branches into mine (resolving a few conflicts that wouldn't have been present if they'd been done all at once), and landing the result.

I expect this to be a common occurrence, especially if we end up doing runs weekly. Renovate will still be saving us time overall, but it's a bit of a project to babysit this process, especially since we run renovate once an hour (self hosted on GitLab SaaS).

It would have been nice if there was a way to tell renovate, "please combine these five PRs into one". I'm not sure what the UX for this should be like. There could be a checkbox in the dashboard issue that says "combine all selected MRs", and on its next it would close those and open a new combined one. That might be gross if you wanted to do that more than once, since there's no obvious way to make more than one group. It occurred to me that you could write a comment that would quote the relevant part of the issue and then the word "combine" beneath it:
```md
>     * [ ]  build(deps): update dependency @sindresorhus/is to v5
> 
>     * [ ]  build(deps): update dependency aggregate-error to v4
> 
>     * [ ]  build(deps): update dependency cacheable-lookup to v7
combine
```
to pick a couple from the current dashboard. You could arrange them into groups however you wanted. Sadly, I don't see the equivalent of GitHub's "Quote reply" issue menu item in GitLab (which is where I want it). I did find that you can highlight the section of text you want to quote and hit `r`, but what it puts into the comment box is a bit ugly:
```md
> 
>
> * [ ] [chore(deps): update golang docker digest to 8e5a006](/.../-/merge_requests/84)
>
>
> * [ ] Convert to task
>   * Delete
> * [ ] [chore(deps): update test-common docker tag to v0.1.17](/.../-/merge_requests/85)
> * [ ] Convert to task
>   * Delete
> * [ ] [chore(deps): update module github.com/pressly/goose/v3 to v3.13.4](/.../-/merge_requests/86)
```
but maybe it gets the job done. Assuming this is of any interest at all.

**Comment by HonkingGoose** [comment]
> But I would be interested in a way to do this on the fly. For instance, the initial set of PRs after onboarding for one of my repos were mostly trivial, and didn't need to land individually, triggering O(n^2) rebuilds as the rest got rebased. I ended up creating a new branch, merging renovate's branches into mine (resolving a few conflicts that wouldn't have been present if they'd been done all at once), and landing the result.

Sounds like having #5573 would help you. You could put those PRs in a Merge Train, and let GitLab merge it.

> It would have been nice if there was a way to tell renovate, "please combine these five PRs into one". I'm not sure what the UX for this should be like. There could be a checkbox in the dashboard issue that says "combine all selected MRs", and on its next it would close those and open a new combined one. That might be gross if you wanted to do that more than once, since there's no obvious way to make more than one group. It occurred to me that you could write a comment that would quote the relevant part of the issue and then the word "combine" beneath it:

We also have issue #4404 about rolling-up multiple updates into a single PR.

I think editing the text of the Dependency Dashboard is a bad user-experience. As I understand, Renovate expects the user to only _select_ checkboxes, not _edit_ the text. We recently had a edge-case bug, where a user deleted all the Dependency Dashboard text:

- #19563

I expect letting users edit the text will lead to more weird cases like that. Also what if Renovate and the user edit the dashboard at the same time (race condition).

I'll let the maintainers decide what to do with your ideas. 😉

**Comment by dhduvall** [comment]
> Sounds like having #5573 would help you. You could put those PRs in a Merge Train, and let GitLab merge it.

Merge trains have a number of restrictions. Fast-forward merge requests are the reason they wouldn't work for us, but they're also restricted to premium GitLab subscriptions. And it leaves GitHub users in the cold, too.

> > It would have been nice if there was a way to tell renovate, "please combine these five PRs into one". I'm not sure what the UX for this should be like. There could be a checkbox in the dashboard issue that says "combine all selected MRs", and on its next it would close those and open a new combined one. That might be gross if you wanted to do that more than once, since there's no obvious way to make more than one group. It occurred to me that you could write a comment that would quote the relevant part of the issue and then the word "combine" beneath it:
> 
> We also have issue #4404 about rolling-up multiple updates into a single PR.

But that also requires renovate configuration changes, which means knowing the possible combinations in advance. I don't want this to be all or nothing, either.

> I think editing the text of the Dependency Dashboard is a bad user-experience.'

I can't disagree, but that's not what I said: "write a comment that would quote the relevant part of the issue". I wouldn't touch the issue description at all.

**Reply by HonkingGoose** [reply]
> > Sounds like having #5573 would help you. You could put those PRs in a Merge Train, and let GitLab merge it.
> 
> Merge trains have a number of restrictions. Fast-forward merge requests are the reason they wouldn't work for us, but they're also restricted to premium GitLab subscriptions. And it leaves GitHub users in the cold, too.

I don't know enough about GitLab's Merge Trains to respond to your whole comment. I just want to say that GitHub has Merge Queues, which is similar. Organizations on GitHub with public repositories can use that for free. And Merge Queues work with Renovate. [^1]

> > I think editing the text of the Dependency Dashboard is a bad user-experience.'
> 
> I can't disagree, but that's not what I said: "write a comment that would quote the relevant part of the issue". I wouldn't touch the issue description at all.

I'm sorry! I misunderstood your workflow... So to check, your proposed workflow is:

1. Go to Dependency Dashboard issue.
2. Go into the edit-post-mode, copy the Markdown checkboxes for the updates you want.
3. Copy/paste the Markdown into a new comment.
4. Edit the copied Markdown so the checkboxes have a tick: `- [x]`
5. Renovate reads the comment, and sees the ticked checkboxes.
6. Renovate creates a merge request with all the updates from the comment in it.

The user experience is _slightly_ better than editing the Dependency Dashboard issue body directly. At least it avoids a potential race condition, where Renovate and user edit the issue body text at the same time.

[^1]: https://docs.renovatebot.com/key-concepts/automerge/#github-merge-queue

**Reply by dhduvall** [reply]
> > > Sounds like having #5573 would help you. You could put those PRs in a Merge Train, and let GitLab merge it.
> > 
> > 
> > Merge trains have a number of restrictions. Fast-forward merge requests are the reason they wouldn't work for us, but they're also restricted to premium GitLab subscriptions. And it leaves GitHub users in the cold, too.
> 
> I don't know enough about GitLab's Merge Trains to respond to your whole comment. I just want to say that GitHub has Merge Queues, which is similar. Organizations on GitHub with public repositories can use that for free. And Merge Queues work with Renovate.

I don't know much about either, so maybe this is a good answer for GH, but not for GL (and I've no clue about other systems).

> > > I think editing the text of the Dependency Dashboard is a bad user-experience.'
> > 
> > 
> > I can't disagree, but that's not what I said: "write a comment that would quote the relevant part of the issue". I wouldn't touch the issue description at all.
> 
> I'm sorry! I misunderstood your workflow...

No worries—it's kinda gross however you look at it. 😄 

> So to check, your proposed workflow is:
> 
>     1. Go to Dependency Dashboard issue.
> 
>     2. Go into the edit-post-mode, copy the Markdown checkboxes for the updates you want.
> 
>     3. Copy/paste the Markdown into a new comment.

This would have that effect, though you can combine 2 & 3 by just highlighting the portion of the issue you want to quote and hit `r`. This will put the markdown of the highlighted text into a new comment. This works for GitLab (for me), but is kind of ugly (my second code block in the OP, but while GitHub says it does this, it doesn't work for me.

Or just highlight the lines and use normal copy/paste, then convert the pasted block into a quoted block. On GitHub, that's: highlight; Cmd-C; scroll; click; Cmd-V; highlight; Cmd-Shift-`.`. If it's the only thing in the comment, then highlighting is just one key-combination, too. A little less clickety-click. Not quite as nice on GitLab.

>     4. Edit the copied Markdown so the checkboxes have a tick: `- [x]`

No need. Just add a keyword (like `combine`) below the list of PRs you want to combine (see my first code block in the OP). If you have multiple quoted sections, each with `combine` after it, then the PRs in each section get combined.

>     5. Renovate reads the comment, and sees the ticked checkboxes.
> 
>     6. Renovate creates a merge request with all the updates from the comment in it.
> 
> 
> The user experience is _slightly_ better than editing the Dependency Dashboard issue body directly. At least it avoids a potential race condition, where Renovate and user edit the issue body text at the same time.

Yeah, I'm definitely not saying it's great. I was just casting a net for some solution that could be implemented in the fairly narrow confines of issue editing that these platforms give us. Maybe the UX design is sufficiently ugly that this basically just can't be done. Not the first time I've wished for the ability to enhance GitLab's UI.

**Reply by HonkingGoose** [reply]
@rarkins what do you think of the ideas in this Discussion?
